File: iosApp/iosApp/Modules/Account ✅/AccountWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultAccountWireframe {
    private weak var parent: UIViewController?
    private let context: AccountWireframeContext
    private let currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory
    private let currencySendWireframeFactory: CurrencyCurrencyWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let deepLinkHandler: DeepLinkHandler
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    private let walletService: WalletService
    private let etherScanService: EtherScanService
    private let settingsService: SettingsService

    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: AccountWireframeContext,
        currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
        currencySendWireframeFactory: CurrencyCurrencyWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        deepLinkHandler: DeepLinkHandler,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService,
        walletService: WalletService,
        etherScanService: EtherScanService,
        settingsService: SettingsService
    ) {
        self.parent = parent
        self.context = context
        self.currencyReceiveWireframeFactory = currencyReceiveWireframeFactory
        self.currencySendWireframeFactory = currencySendWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.deepLinkHandler = deepLinkHandler
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
        self.walletService = walletService
        self.etherScanService = etherScanService
        self.settingsService = settingsService
    }
}

extension DefaultAccountWireframe {
    
    func present() {
        let vc = wireUp()
        self.vc = vc
        print("[DefaultAccountWireframe] Presenting")
        if settingsService.isSelected(
            setting: .init(group: .developer, action: .developerTransitionsSheet)
        ) {
            vc.modalPresentationStyle = .automatic
            parent?.show(vc, sender: self)
        } else {
            let presentedTopVc = (vc as? UINavigationController)?.topVc
            let delegate = presentedTopVc as? UIViewControllerTransitioningDelegate
            vc.modalPresentationStyle = .custom
            vc.transitioningDelegate = delegate
            parent?.present(vc, animated: true)
        }
    }

    func navigate(with destination: AccountWireframeDestination) {
        if destination is AccountWireframeDestination.Receive {
            let context = CurrencyReceiveWireframeContext(network: context.network, currency: context.currency)
            currencyReceiveWireframeFactory.make(vc, context: context).present()
        }
        if destination is AccountWireframeDestination.Send {
            let context = CurrencySendWireframeContext(
                network: context.network,
                address: nil,
                currency: context.currency
            )
            currencySendWireframeFactory.make(vc, context: context).present()
        }
        if destination is AccountWireframeDestination.Swap {
            let context = CurrencySwapWireframeContext(
                network: context.network,
                currencyFrom: context.currency,
                currencyTo: nil
            )
            currencySwapWireframeFactory.make(vc, context: context).present()
        }
        if destination is AccountWireframeDestination.More {
            deepLinkHandler.handle(deepLink: .degen)
        }
    }
}

private extension DefaultAccountWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultAccountInteractor(
            currencyStoreService: currencyStoreService,
            walletService: walletService,
            etherScanService: etherScanService
        )
        let vc: AccountViewController = UIStoryboard(.account).instantiate()
        let presenter = DefaultAccountPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.edgesForExtendedLayout = []
        vc.presenter = presenter
        return NavigationController(rootViewController: vc)
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/account/AccountWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.account

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive.CurrencyReceiveWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencysend.CurrencySendWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.account.*
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySend.CurrencySendWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapWireframeContext
import com.sonsofcrypto.web3walletcore.services.etherScan.EtherScanService

class DefaultAccountWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: AccountWireframeContext,
    private val currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
    private val currencySendWireframeFactory: CurrencySendWireframeFactory,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val currencyStoreService: CurrencyStoreService,
    private val walletService: WalletService,
    private val etherScanService: EtherScanService,
): AccountWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.get()?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: AccountWireframeDestination) {
        when (destination) {
            is AccountWireframeDestination.Receive -> {
                val context = CurrencyReceiveWireframeContext(context.network, context.currency)
                currencyReceiveWireframeFactory.make(parent?.get(), context).present()
            }
            is AccountWireframeDestination.Send -> {
                val context = CurrencySendWireframeContext(context.network, null, context.currency)
                currencySendWireframeFactory.make(parent?.get(), context).present()
            }
            is AccountWireframeDestination.Swap -> {
                val context = CurrencySwapWireframeContext(context.network, context.currency, null)
                currencySwapWireframeFactory.make(parent?.get(), context).present()
            }
            is AccountWireframeDestination.More -> {
                println("[AA] Navigate to More...(deep link)")
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = AccountFragment()
        val interactor = DefaultAccountInteractor(
            currencyStoreService,
            walletService,
            etherScanService,
        )
        val presenter = DefaultAccountPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}